CPURegion allocator

ajyu opened the pull request:
Summary: New allocator that mimics behavior of BlackBoxPredictor's buffer

Differential Revision: D10785119

ezyang commented on aten/src/ATen/core/Allocator.h:
```diff
@@ -119,6 +119,12 @@ struct Allocator {
     AT_ASSERT(d);
     d(ptr);
   }
+  virtual void prepare(size_t s) {
```
These need docs. What are they supposed to do? Why do they have to be virtual methods? (Note that you're always free to define extra methods on the subclass and call them directly; it's only necessary to put something as a virtual method on the parent class if you need to call them in a situation where you **don't** know what the concrete implementation is. So maybe you don't even need these at all!)

ezyang commented:
We should write an OSS tests for this, so that it can be run outside of FB.

ajyu committed:
free blobs after memonger

Differential Revision: D12840803

fbshipit-source-id: cfbea708ab77e99829325fb792098100754a73a8

ajyu committed:
CPURegion allocator (#13077)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/13077

Allocator that mimics behavior of BlackBoxPredictor's buffer: Allocates a bigger chunk of memory than needed, then returns addresses from that buffer serially. If we run out, allocate another chunk.

What this does not do: 1) This does not actually replace BBP's buffer. We plan to do that after this allocator is well vetted and replaces the functionality of the current DefaultCPUAllocator. 2) This does not turn on "free by default" (to be also tackled in a future diff).

More context in https://fb.quip.com/r64bAyyReNmk

Differential Revision: D10785119

fbshipit-source-id: 2a316b64697fc98396a6d08c79db70ab8f499c24

ezyang commented on c10/core/Allocator.h:
```diff
@@ -127,6 +127,10 @@ struct C10_API Allocator {
+  // Used by allocators that maintain state, i.e. CPURegionAllocator
+  virtual void reset() {
+    return;
+  }
```
I'm a bit ambivalent about this. Under the current allocator model, all `Allocator` objects are expected to be static lifetime. That implies that there's no need to actually put the state in the Allocator object itself, and generally this is a good thing because the Allocator interface exists purely to let us dynamic dispatch to the allocate function when we don't know what the allocator is; but interesting allocators have all sorts of extra knobs, and you don't want to add all of those knobs to Allocator.

The key litmus test is, will you ever call `reset()` on a generic `Allocator` which you don't know what it is? In the non-OSS portion of this patch, there is a call to reset(), but in a context where we *know* the region allocator is being used. That to me is a strong indicator we shouldn't have this virtual method on this class.

ezyang replied:
OK, looking at CPURegionAllocator, I see that you do, in fact, want to have multiple instances of the Allocator object. So maybe we should support this pattern, but we will have to be careful about how heavily we've made assumptions about this elsewhere. One particular hazard I can think of right off the bat is that if you have a Tensor allocated by CPURegionAllocator, and it persists beyond the scope of the CPURegionAllocator, it will end up with a dangling Allocator pointer (as well as dangling memory, so maybe this is not the... worst thing in the world? But it is pretty nasty for users who expect tensors to stay valid anywhere.) cc @Yangqing who I remember advocated quite strongly for static lifetime allocators.

ezyang commented:
cc @jerryzh168 @dzhulgakov  this affects allocator unification

ezyang commented on caffe2/core/allocator.h:
```diff
@@ -42,6 +43,8 @@ class CAFFE2_API MemoryAllocationReporter {
   std::unordered_map<void*, size_t> size_table_;
   size_t allocated_;
+  size_t allocations_;
+  size_t deallocations_;
```
I didn't see you initialize these to zero anywhere in the patch; they need to be explicitly initialized (I recommend putting `= 0` here)
